### PR TITLE
Convert RAW vectors to Python bytearray; Convert Python bytearray to RAW

### PR DESCRIPTION
**Pull request by jjallaire**
Addresses https://github.com/rstudio/reticulate/issues/303

@kevinushey Could you review?

Trying to think through whether we should merge this now (CRAN release process will start in ~ 7 days) or wait. The only changes in behavior are:

- RAW is now always converted to bytearray (previously it was string for Python 2 which is plain wrong and bytes for Python 3). 

- bytearray is now converted to RAW (previously it had no conversion)

I considered whether we should convert bytes to RAW, however I think we can't do this for Python 2 (where bytes is the same thing as string) and I don't think we should introduce behavior that differs across Python 2 and Python 3.

### CODE REVIEW

**Comment by jeffwong-nflx** [comment]
Thanks JJ! This is really helpful for me. I'm working on a use case where I serialize bytes from R to storage. There is a python client, motivating the `serialize() %>% r_to_py` call, and to retrieve the bytes from storage the opposite `py_to_r %>% unserialize` call

File: R/python.R [hunk 1]
```diff
@@ -981,6 +981,15 @@ py_str.python.builtin.object <- function(object, ...) {
   str
 }
 
+#' @export
+py_str.python.builtin.bytearray <- function(object, ...) {
+  builtins <- import_builtins()
+  paste(
+    paste0("python.builtin.bytearray (length ", builtins$len(object), "):"),
+    capture.output(str(py_to_r(ba), give.length = FALSE, give.head = FALSE, vec.len = 10))
```
Comment (kevinushey) [inline R/python.R]: Looks like that `ba` should be `object`?

File: src/python.cpp [hunk 2]
```diff
@@ -1167,7 +1167,8 @@ PyObject* r_to_py_cpp(RObject x, bool convert) {
   // bytes
   } else if (type == RAWSXP) {
 
-    return as_python_bytes(RAW(sexp), Rf_length(sexp));
+    Rcpp::RawVector raw(sexp);
+    return PyByteArray_FromStringAndSize((const char*)&raw[0], raw.size());
```
Comment (kevinushey) [inline src/python.cpp]: This is pedantic but `&raw[0]` could technically be an out-of-bounds access for an empty vector.

File: R/python.R (hunk 1 above)
Comment (jjallaire) [reply R/python.R]: Good catch! I actually just changed this to just print the number of bytes and the type (as the initial implementation actually converted the bytearray to RAW which is probably too expensive for a print method)

**Comment by jeffwong-nflx** [comment]
In this PR could we also address another issue?

[When reticulate converts RAW to bytes](https://github.com/rstudio/reticulate/blob/91ba73dd457ce111cb58631893d7b148e1a3b328/src/python.cpp#L1170), it measures the size of RAW using Rf_length(). This is limited to sizes of 2GB. An alternative that allows longs is Rf_xlength. See this issue from [Rcpp](https://github.com/RcppCore/Rcpp/issues/302)

Without this, a call to serialize %>% r_to_py will throw an error if the bytes are too large

File: src/python.cpp (hunk 2 above)
Comment (jjallaire) [reply src/python.cpp]: Not pedantic at all as that's a real failure mode in both directions! (thanks for catching that). Just pushed a change, could you take a quick look (wrote tests but another pair of eyes always useful)

**Comment by jeffwong-nflx** [comment]
I checked out your branch and everything worked well for my use case. Thanks!

**Comment by jjallaire** [comment]
@jeffwong-nflx Thanks for verifying!

@kevinushey You can merge this after final review.

**Pull request merged**